[botosek] iP

botosek opened the pull request:
### MomBot ###

> "Now, MomBot nags at me more than my parents do" -someone, probably.😬

MomBot frees your mind. It lets you be a person your parents would finally approve of. It's,
- Text-based
- Easy to use

Using it is easy:
1. Enter your task with the todo/deadline/event command
2. Include the specified parameters if necessary
3. And you are **done**!

Features:
- [x] Managing Tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

click [here](url) to download it for yourself now!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

botosek committed:
Implement initial skeletal code

botosek committed:
Include command echoing

botosek committed:
Add ability to store text entered by user as tasks and display them back when requested

botosek committed:
Add ability to mark and unmark tasks as done and undone respectively

botosek committed:
Add support to track 3 types of tasks(Todo, Event, Deadline)

botosek committed:
Use the input/output redirection technique to semi-automate testing of chatbot

botosek committed:
Add error handling for invalid commands and invalid inputs of todo commands

botosek committed:
Add support for deleting tasks from the list

botosek committed:
Use Java enums for user commands

botosek committed:
Add support for saving task list to hard drive

botosek committed:
Add support for files with corrupted data

botosek committed:
Merge branch-Level-7 back to master

botosek committed:
Add task datetime functionality

botosek committed:
Merge branch-Level-8 back to master

botosek committed:
Modify main code to become more OOP

botosek committed:
Modify Storage code to be more OOP

botosek committed:
Merge branch-A-moreOOP back to master

botosek committed:
Order class files into packages

botosek committed:
Merge A-packages back to master

botosek committed:
Merge add-gradle-support back to master

botosek committed:
Add unit tests for task strings saved to hard disk and invalid commands

botosek committed:
Merge branch-A-JUnit back to master

botosek committed:
Modify file access method to make data folder open in the same directory as packaged jar file

botosek committed:
Merge branch-A-Jar back to master

botosek committed:
Clean program directory

botosek committed:
Add find feature

botosek committed:
Modify code to adhere to coding standard

botosek committed:
Add JavaDoc comments

botosek committed:
Merge branch-A-JavaDoc back to master

botosek committed:
Merge branch-A-CodingStandard back to master

botosek committed:
Merge branch-Level-9 back to master

botosek committed:
Add checkstyle

botosek committed:
Merge branch-A-CheckStyle back to master

botosek committed:
Add graphic user interface

Melvan-Chan commented on src/main/java/mom/Mom.java:
```diff
@@ -0,0 +1,88 @@
+ */
+public class Mom implements Parser {
+    /** Relative file path of hard disk. */
+    private static final String TASKLIST_FILEPATH = "data" + File.separator + "mom.txt";
```
I like the use of all-caps noun for a constant, following Java conventions.

Melvan-Chan commented on src/main/java/mom/Mom.java:
```diff
@@ -0,0 +1,88 @@
+     * @param filePath relative file path of hard disk file.
+     */
+    public Mom(String filePath) {
+        ui = new Ui(CHATBOT_NAME);
```
Similar to above, good use of an all-caps noun, and it's descriptive.

Melvan-Chan commented on src/main/java/mom/Mom.java:
```diff
@@ -0,0 +1,88 @@
+     * Runs the main program.
+     */
+    public void run() {
+        ui.displayIntro();
```
I like the use of a verb for the method.

Melvan-Chan commented on src/main/java/mom/Parser.java:
```diff
@@ -0,0 +1,197 @@
+     * @param  userCommand String taken from user input.
+     * @return             A boolean indicating if the string is a valid command.
+     */
+    static boolean checkValidCommand(String userCommand) {
```
I like the use of a verb for the method to indicate the action of checking if the command is valid. Good choice.

Melvan-Chan commented on src/main/java/mom/TaskList.java:
```diff
@@ -0,0 +1,212 @@
+     * @param offset   The offset number where the description starts in the raw user input string.
+     * @throws InvalidInputException If the raw user input is not a valid command.
+     */
+    public static void handleTask(Ui ui, TaskList taskList, Command command, String input, int offset)
```
I like the simple namings of methods. Descriptive enough of what they do.

Melvan-Chan commented on src/main/java/mom/Ui.java:
```diff
@@ -0,0 +1,141 @@
+    /**
+     * Provides shorter function name to display a message to user.
+     */
+    public static void display(String message) {
```
I like the method name for it being clear and concise, and for the other display methods which has a descriptive verb following the word "display".

Melvan-Chan reviewed (commented):
Good discipline with the use of nouns for variables and verbs for methods.

botosek committed:
Merge branch-Level-10 back to master

botosek committed:
Remove some redundant code

jordon0802 commented on data/mom.txt:
```diff
@@ -0,0 +1,6 @@
+D | 0 | return book | 6/6/2025 0000
+D | 1 | eat bao | 3/12/2025 0000
+E | 0 | pop balloons | 12/4/2025 1800-13/4/2025 0600
+T | 0 | read app
+E | 0 | help bob | 6/5/2012 0000-7/6/2040 0000
+T | 1 | complete work
```
Perhaps storage file should not be tracked?

jordon0802 commented on src/main/java/mom/Storage.java:
```diff
@@ -0,0 +1,138 @@
+    public void handleFileEntries(String[] entryList, String entry) throws CorruptedFileException {
+        String commandString = entryList[0];
+        switch (commandString) {
+        case "T": {
```
Curly braces for each switch case can perhaps be removed?

jordon0802 commented on src/main/java/mom/command/Command.java:
```diff
@@ -0,0 +1,8 @@
+/*
+ * Command data type containing all valid commands for the chatbot.
+ */
+public enum Command {
+    regular, bye, list, mark, unmark, delete, todo, deadline, event, find
+}
```
Should enum constants be capitalised?

jordon0802 commented on src/main/java/mom/Storage.java:
```diff
@@ -0,0 +1,138 @@
+        this.filePath = getFilePath(filePath);
+    }
+
+    /**
+     * If it is run as a packaged .jar file, looks for or creates hard disk file in reference to project root.
+     * Else, looks for or creates it reference to the project root.
+     *
+     * @param filePath File path of the hard disk file in reference to the program file.
+     * @return File path of the hard disk file to be loaded from or saved into.
+     */
+    public static String getFilePath(String filePath) {
+        File jarFile = new File(Mom.class.getProtectionDomain().getCodeSource().getLocation().getPath());
+        File f = new File(jarFile.getParentFile().getParent(), filePath);
+
+        if (jarFile.toString().endsWith(".jar")) {
+            f.getParentFile().mkdirs();
+            return f.getAbsolutePath();
+        } else {
+            return "." + File.separator + filePath;
+        }
+    }
```
LGTM!! Like that you considered the app running from a jar package!

jordon0802 reviewed (commented):
Overall, code looks really well-structured! Just some minor nits😬

botosek committed:
Add assert statements

The code currently does not contain any assert statements and relies
on exceptions to detect errors. Some errors might thus be left
undetected.

Adding some assertion statements can help verify assumptions
made in the code.

Let's add some assert statements to check the return type of some
of the functions in the TaskList class.

botosek committed:
Adjust code quality

There are some minor errors in the code that do not meet the code
quality standard. There are also some functions that are missing
Javadoc comments.

botosek committed:
Merge pull request #1 from botosek/branch-A-Assertions

Add assert statements

botosek committed:
Merge pull request #2 from botosek/branch-A-CodeQuality

Adjust code quality

botosek committed:
Add JavaDoc to methods in Ui class

Some methods in the Ui class had JavaDocs that were insufficient in
explaining the purpose, parameters and return values and types of the
methods.

Expanding them and ensuring that they meet and all conform to the
right standard makes the code easier to understand.

Let's provide further elaboration and include the relevant
parameter and return variables in the JavaDocs.

botosek committed:
Include undo functionality

There is no way for the user to revert a command if they have made a
mistake.

To allow the user to revert to a previous stage, an "undo" feature was
added, by putting the different states into an array list and adjusting
the pointer accordingly if an undo command is called.

This allows the different states of the tasklist to be saved easily.

botosek committed:
Merge branch-C-Undo back to master

botosek committed:
Add main window name

botosek committed:
Fix mark and unmark task logic

botosek committed:
Add User Guide

botosek committed:
Merge branch-A-UserGuide back to master

botosek committed:
Clean user guide

botosek committed:
Fix pictures in user guide

botosek committed:
Attempt fix on pictures in user guide

botosek committed:
Fix command table in user guide

botosek committed:
Attempt fix for command table in user guide

botosek committed:
Modify undo command response

botosek committed:
Attempt fix on table in showcase page

botosek committed:
Modify main README.md file

botosek committed:
Fix Ui.png image
